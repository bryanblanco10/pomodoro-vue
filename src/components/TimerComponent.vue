<template>
	<section>{{minutes}} : {{seconds}}</section>
</template>
<script>
	export default{
		name: 'TimerComponent',
		data(){
			return{
				timeInSeconds: 0,
				newTimeSeconds: ''
			}
		},
		props:{
			time:{
				type: Number,
				default: 1200
			}
		},
		watch:{
			time(newTime){
				this.newTimeSeconds = newTime
			}
		},
		computed:{
			minutes(){
				const minutes = Math.floor(this.newTimeSeconds / 60) 			
				return this.getPaddedTime(minutes)
			},
			seconds(){
				// % devuelve el resto
				const seconds = Math.floor(this.newTimeSeconds % 60) 
				return  this.getPaddedTime(seconds)
			}
		},
		mounted(){
			this.newTimeSeconds = this.$props.time
			this.startCountdown()
		},
		methods:{
			startCountdown(){
				const countdownTimer = window.setInterval(() => {
					if(!this.newTimeSeconds){
						window.clearInterval(countdownTimer)
						return
					}
					this.onSecondElapse()
				}, 1000)
			},
			onSecondElapse(){
				this.newTimeSeconds--
			},
			getPaddedTime(value){
				return value.toString().padStart(2, '0')
			}
		}
	}
</script>
<style type="text/css" scoped>
	
</style>