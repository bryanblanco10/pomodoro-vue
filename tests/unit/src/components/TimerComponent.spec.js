import { shallowMount, createLocalVue } from '@vue/test-utils'
import TimerComponent from '@/components/TimerComponent.vue'

jest.useFakeTimers()

const localVue = createLocalVue()

function createTimerComponent(propsData = {time: 121}){
	 return shallowMount(TimerComponent,{
		propsData,
		localVue
	})
}

//Para saber que se esta renderizando bien.
describe('TimerComponent.vue', () => {
  it('should render property', (done) => {
   	const timer = createTimerComponent()

    timer.vm.$nextTick(() =>{
   	 expect(timer).toMatchSnapshot()
   	 done()
    })
  })
})

describe('Computed',() =>{
	it('should return minutes property and padded if necessary', () =>{
		const timer = createTimerComponent()

   	expect(timer.vm.minutes).toEqual('02')
	})

	it('should return seconds property and padded if necessary', () =>{
		const timer = createTimerComponent()

   	expect(timer.vm.seconds).toEqual('01')
	})
})

describe('Methods', () =>{
	describe('startCountdown', () =>{
		it('should call onSecondElapse when a second passes', () =>{
			const timer = createTimerComponent()
			const onSecondElapse = jest.spyOn(timer.vm,'onSecondElapse')
			jest.advanceTimersByTime(2000)
			expect(onSecondElapse).toHaveBeenCalledTimes(2)
		})

		it('should clear interval when time is 0', () =>{
			const timer = createTimerComponent({time: 1})
			const onSecondElapse = jest.spyOn(timer.vm,'onSecondElapse')
			jest.advanceTimersByTime(2000)
			expect(onSecondElapse).toHaveBeenCalledTimes(1)
		})
	})

	describe('onSecondElapse', () =>{
		it('should set a second less to timeInSeconds', () =>{
			const seconds = 3
			const timer = createTimerComponent({time: seconds})
			timer.vm.onSecondElapse()

			expect(timer.vm.newTimeSeconds).toEqual(seconds - 1)
		})
	})

	describe('getPaddedTime', () => {
		it('should return number value padded with 0', () =>{
			const seconds = 5
			const timer = createTimerComponent({time: seconds})

			expect(timer.vm.getPaddedTime(5)).toEqual('05')
		})
	})
})

// describe('Watch', () =>{
// 	describe('time',() =>{
// 		it('should call time when time property change', () =>{
// 			const timer = createTimerComponent()
// 			expect(timer.vm.timeInSeconds).toEqual(121)

// 			timer.setProps({
// 				time: 100
// 			})

// 			expect(timer.vm.timeInSeconds).toEqual(100)
// 		})
// 	})
// })
