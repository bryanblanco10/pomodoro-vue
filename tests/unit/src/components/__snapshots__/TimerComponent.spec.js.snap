// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimerComponent.vue should render property 1`] = `<section>02 : 01</section>`;
